moderntribe/wme
# Nexcess domains

I have been working on adding support o search domains using Nexcess. 

## Problem
Until now, searching for available domains within the app was impossible. Instead, the app directly redirected you to Nexcess for the whole process.

## Solution
To streamline the process, domain search and selection have been integrated into the go-live wizard. To this end, new screens have been created based on [the designs](https://www.figma.com/file/1poWoDLfFnVcFrRH3wIDyC/WME-Core?node-id=1134%3A23261).
- Find Domain. It shows a list of domains based on text input and allows you to select a domain.
- Connect with Nexcess. It shows the selected domain and a button to connect with Nexcess
- Claim Your Domain. It shows a success page

When a user clicks on "Connect Your Nexcess Account", the app makes a request to the backend that returns an URL. After receiving it, the app redirects the user to Nexcess for checkout.

After finishing the checkout process, Nexcess will redirect the user back to the app. This URL should be: `/wizard/go-live?nexcess=true&step=3&domain=jorgetri.be`. Search params are important:
- `nexcess=true` and `step=3` are used to know the state of the routes
- `domain=jorgetri.be` is used to show the user the bought domain

I have introduced some styles, but most have been reused from other existing components.

I installed [**react-query**](https://tanstack.com/query/v4/docs/adapters/react-query) to help us manage the state of the queries and mutations. It is much more powerful, but in our case, it helps us in 2 obvious cases:
- It reduces boilerplate (and thus errors) to handle the state of queries: `isLoading`, `isError`, etc.
- It cache in the client the previous queries. This gives a better UX. If, in the same session, you search for jorge.com, then change to jorge2.com and then search for jorge.com again. We will immediately show you the cached results, as 99.9% of the cases will be valid. In case of a change, they will be replaced with new data.

### Screenshots
![Screenshot 2022-10-28 at 10 57 52](https://user-images.githubusercontent.com/28598593/198562994-199c9bd1-c88e-4910-bb9a-0d5a4c56f327.png)
![Screenshot 2022-10-28 at 10 58 01](https://user-images.githubusercontent.com/28598593/198563021-56c9238c-df1d-4a3b-b504-e8e4ad0db19c.png)
![Screenshot 2022-10-28 at 11 10 20](https://user-images.githubusercontent.com/28598593/198563044-ba685d11-b763-42c7-ba68-74f8dd031885.png)


## How to test
These changes must be deployed in the WordPress plugin.

1. Open the sitebuilder
2. Open the "Go Live" wizard
3. Select "No, I need one!"
4. Search for a domain
   - the domain state can be: taken, available or selected
   - only one domain can be selected at a time
5. Select and domain and click continue
6. See a summary of the selected domain
7. Click on "Connect Your Nexcess Account"
   - It should redirect you to nexcess for checkout
   - And once the checkout is completed, be redirected to a success page in our app
8. Click Save & Exit
   - Be directed to the sitebuilder homepage
